app/services/v1/question/answer_options_change_detector.rb

```diff
@@ -0,0 +1,484 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ClassLength
+class V1::Question::AnswerOptionsChangeDetector
+  attr_reader :question
+
+  def initialize(question)
+    @question = question
+  end
+
+  def detect_changes(old_options, new_options)
+    case question
+    when ::Question::Single
+      result = compute_single_changes(old_options, new_options)
+    when ::Question::Multiple
+      result = compute_multiple_changes(old_options, new_options)
+    when ::Question::Grid
+      result = compute_grid_row_changes(old_options, new_options)
+    else
+      result = []
+    end
+    result
```

Comment:
I suggests to extract this logic to separated service by:
a) each function moves to dedicated service `detect_changes -> call new Service`
b) each question types has dedicated service